### DISCUSSION TITLE
got stuck with double quotes in Berry

### DISCUSSION THREAD

**Discussion by indigmas** [arendst/Tasmota#20983] [Q&A] [answered]
Hi community,
i want to do a Modbussend.
In Tasmota it works like this
`Modbussend{"deviceAddress":1, "functionCode":4, "startAddress":3218, "type":"int16","count":1}`
In Berry i tried several ways.
`tasmota.cmd("Modbussend{"deviceAddress":1,"functionCode":4,"startAddress":3218,"type":"int16","count":1}")`
or
`tasmota.cmd("Modbussend{""deviceAddress"":1,""functionCode"":4,""startAddress"":3218,""type"":""int16"",""count"":1}")`
The most reasonable (because referring to the definition in the berry short manual) was this:
`tasmota.cmd("Modbussend{/"deviceAddress/":1,/"functionCode/":4, /"startAddress/":3218, /"type/":/"int16/",/"count/":1}")`
what is wrong? Or is it the completely wrong attempt to make a ModbusSend out of Berry?

**Comment by sfromis** [comment, marked as answer]
The simple approach would be to observe that in Berry you can use single and double quotes the same way. You could also escape either within a string using backslash. To construct JSON, you can also use the already existing module for that.

```
tasmota.cmd('Modbussend {"deviceAddress":1,"functionCode":4,"startAddress":3218,"type":"int16","count":1}')
tasmota.cmd("Modbussend {\"deviceAddress\":1,\"functionCode\":4,\"startAddress\":3218,\"type\":\"int16\",\"count\":1}")
import json
tasmota.cmd("Modbussend " + json.dump({"deviceAddress":1,"functionCode":4,"startAddress":3218,"type":"int16","count":1}'))
```
The last construct does not look like an improvement, the point is that you are creating a map in Berry, and then you can serialize that map as a JSON string. You can handle such a map stepwise, like setting/getting/removing individual fields. That syntax also allows using vars where you have the numbers (and also the strings), as it all is expressions in the Berry language.

The above examples also correct the minor syntax error that you had no space after the command name.

**Reply by indigmas** [reply]
thank you for your very quick response :)
at first glance, i think i'll take the first one, because it looks the simplest way to me ;)

**Reply by indigmas** [reply]
i thought twice - i take the json version, because like you proposed - i need to use some kind of variables :)

**Reply by sfromis** [reply]
If you *literally* mean the command answer, the `tasmota.cmd` function will return that. However as `ModbusReceived` is not from the command, but just incoming data later, this will not be useful.

To process this payload, see the `tasmota.add_rule` examples in the Berry docs (this is *not* adding "classic" rules). You can use triggers in mostly the same format as in rules. However, you are not as badly restricted to only being able to get a single value, but you also get the whole JSON already parsed into a map. Unlike "classic" rules, you can also combine multiple triggers.

The feature of getting the whole JSON into a map is not clearly laid out on the docs page, but could be something like this:
```
tasmota.add_rule('ModbusReceived',
  def (received,trigger,data)
    print(received['Values'][0])
  end
```
In this example `received` is a map for the matched part of the payload (`data` is the whole payload parsed), and hence you can make a map reference to the key `Values`. As your syntax above indicates that you are expecting an array, you'll also get an array in Berry, where the first value is index 0, not index 1 as in rules.
